### ISSUE TITLE
Window functions are very slow

### ISSUE THREAD

**Issue by stas-sl** [duckdb/duckdb#1367]
Window functions are very slow - 5 seconds for 1M rows and a simple query

```python
import time
import random
import duckdb

db = duckdb.connect(':memory:')
db.execute('create table data(id integer, value integer, primary key(id))')

values = []
for i in range(1_000_000):
    values.append((i, random.randint(0, 100000),))
db.executemany('insert into data(id, value) VALUES (?, ?)', values)
db.commit()

start = time.time()
db.execute('select value, row_number() over(order by id) from data limit 1').fetchall()
duration = time.time() - start
print(f'{duration:.3f} s')
```
> `5.367 s`

**Comment by hannesmuehleisen** [comment]
Indeed this could be faster. There are several optimisation opportunities left in the Window function operator. Happy to review  a PR!

**Comment by hawkfish** [comment]
I'd be interested in looking at this. I've worked on the original Hyper Window operator and reimplemented it for the Tableau Data Engine.  Do you have a list of tasks?

**Comment by ankoh** [comment]
We found 3 major pain-points in increasing complexity:
- Windows are very often dominated by sorting and DuckDB currently sorts fully instead of hash partitions.
- BinarySearchRightmost -> ChunkCollection::GetRow and ChunkCollection::Reorder take up a major part of the execution time.
  Both use the Value api and allocate vectors dynamically which really hurts a lot.
- The sorting is always indirect and every comparison resolves the containing chunks (CompareTuple).
  It would go a long way to merge tuples into a single chunk first and then use a vectorized quicksort partitioner.
  (I went for a morsel-driven BlockQuicksort for Umbra which beats an indirect vanilla QuickSort by far https://arxiv.org/abs/1604.06697)

**Comment by hannesmuehleisen** [comment]
So @lnkuiper is currently working on parallelising the sort. We also are going to use this in the Window functions. Within the operator I indeed see the `Value` based operations as the biggest problem.

**Comment by hawkfish** [comment]
At a higher level, there is no attempt made to compare `OVER` specifications that I can see. So if the user has 12 window functions with identical `OVER` specifications (say from a `WINDOW` clause), this slow sorting will be repeated 12 times(!) A simple fix would be to pull the `SortCollectionForWindow` call up to the loop and bin the window functions by `OVER` specification.

**Comment by hawkfish** [comment]
This brings up a larger issue around the current design of the physical window operator.  If you look at [Cao et al.](http://vldb.org/pvldb/vol5/p1244_yucao_vldb2012.pdf) it really suggests that logical operator ordering is determined by the pair `(WPK, WOK)` and multiple "functions" sharing this pair should be optimised as a single unit. If the physical window operator had a single pair like this, then there are multiple opportunities for performance improvement orthogonal to the `Value`-based sorting improvements proposed. In particular, a large amount of pre-processing can be performed on the input threads:
* The sort keys, payloads, frame boundaries and lead/lag parameters can be computed in parallel in `Sink`
* The partitioning keys can then be used immediately to hash these chunks into separate bins on each thread
* `Combine` then merges/appends the chunks in the bins
* `Finalize` can then spawn threads to independently perform the current operations, including sorting as in [Leis et al.](http://www.vldb.org/pvldb/vol8/p1058-leis.pdf). Work stealing can also be implemented using a task queue.
This approach also implies that parallelising the sort code may not be the most effective way to improve sorting performance unless there is only a single bin. At the very least, the sort operation might take a flag indicating whether parallel sorting is appropriate.

**Comment by hawkfish** [comment]
My use case potentially scales to hundreds of millions of rows(!) so I have a brutal test case I can inflict on the code...

**Comment by Mytherin** [comment]
> At a higher level, there is no attempt made to compare OVER specifications that I can see. So if the user has 12 window functions with identical OVER specifications (say from a WINDOW clause), this slow sorting will be repeated 12 times(!) A simple fix would be to pull the SortCollectionForWindow call up to the loop and bin the window functions by OVER specification.

> I ran into this today because I had 6 window queries with two functions (with identical OVERs) and thought that if I made them into a single query it would get faster. But since all the time goes into sorting and the same sort is redone every time, it made little difference.

This is true and an easy optimization opportunity, with potential big pay-off. The current window operator implementation is very naïve in this regard.

> This brings up a larger issue around the current design of the physical window operator. If you look at Cao et al. it really suggests that logical window operator ordering is determined by the pair (WPK, WOK) and multiple "functions" sharing this pair should be optimised as a single unit. If the physical window operator had a single pair like this, then there are multiple opportunities for performance improvement orthogonal to the Value-based sorting improvements proposed. In particular, a large amount of pre-processing can be performed on the input threads:

That is very true, a more efficient sort would help somewhat but indeed specialized optimizations for the window operator seem like the way to go to maximize performance there, particularly taking advantage of the partitioning schemes as you mentioned.

> My use case potentially scales to hundreds of millions of rows(!) so I have a brutal test case I can inflict on the code...

Excellent :) Always nice to have real-world test cases.

**Comment by hawkfish** [comment]
I was thinking that we could replace both calls to `EqualsSubset` and one of the two `BinarySearchRightmost` by just building two bitmaps (one for partitioning and one for ordering) that are 1 when a row is equal to the previous row. The partition bitmap starts as all 1s except for the first value. The order bitmap starts as a copy of the partition bitmap. The two calls to `EqualsSubset` are reduced to bit tests and the first `BinarySearchRightmost` is a bit scan.

Constructing the bitmaps is column-wise so it should be easy to templatise using the physical types. to check for equality.

The second `BinarySearchRightmost` seems to do too much work since it is guaranteed to be within the partition already, so there is no need to compare the partitioning columns? The ordering bitmap could then be used to determine the next  peer boundary.

**Comment by hawkfish** [comment]
Masks improved the OP test by 5x on my MacBook Pro.

**Comment by hawkfish** [comment]
How can I schedule `Task`s in `PhysicalWindow::Finalize`? I want to compute each partition of each window function for a fixed `(WPK, WOK)` pair to run in parallel, but it is not at all clear how this should be done. If I `Schedule` the tasks and then pull them out one at a time, the tasks run but then `Executor::Initialize` hangs waiting for some other task that is never coming.

**Comment by Mytherin** [comment]
The [aggregate HT](https://github.com/cwida/duckdb/blob/master/src/execution/operator/aggregate/physical_hash_aggregate.cpp#L335) has code for this. Essentially you add tasks to the current pipeline by increasing the "total_tasks", and then scheduling tasks using TaskScheduler::ScheduleTask. We should clean this up and create an actual API for this, but for now this is how it can be done.

**Comment by hawkfish** [comment]
I just tried that and it hung. The code looks like:

```
	void Execute() override {
		ComputeWindowExpression(wexpr, input, output, partition_mask, order_mask, output_col, range);
		++parent.finished_tasks;
	}
...
		pipeline.IncrementTasks(matching.size() * ranges.size());
		auto &scheduler = TaskScheduler::GetScheduler(context);

		//	Compute the functions with matching sorts
		for (const auto &expr_idx : matching) {
			D_ASSERT(select_list[expr_idx]->GetExpressionClass() == ExpressionClass::BOUND_WINDOW);
			auto wexpr = reinterpret_cast<BoundWindowExpression *>(select_list[expr_idx].get());
			// reuse partition and order clause in window def
			for (const auto &range : ranges) {
				auto new_task = make_unique<ComputeWindowExpressionTask>(pipeline, wexpr, big_data, window_results, partition_mask, order_mask, expr_idx, range);
				scheduler.ScheduleTask(pipeline.token, move(new_task));
			}
		}

		//	Wait for the tasks to complete before moving on.
		while (pipeline.finished_tasks < pipeline.total_tasks) {
			unique_ptr<Task> 	task;
			if (scheduler.GetTaskFromProducer(pipeline.token, task)) {
				task->Execute();
				task.reset();
			}
		}
```cpp

**Comment by Mytherin** [comment]
Execute in the aggregate task explicitly calls ::Finish on the pipeline when all tasks are completed. Perhaps that is missing?

```cpp
	void Execute() override {
		FinalizeHT(state, radix);
		lock_guard<mutex> glock(state.lock);
		parent.finished_tasks++;
		// finish the whole pipeline
		if (parent.total_tasks == parent.finished_tasks) {
			parent.Finish();
		}
	}
```

**Comment by hawkfish** [comment]
Yes I saw that but the problem is that there is no way to know ahead of time how many `Tasks will be added because each `(WPK, WOK)` pair may generate completely different partitioning.

**Comment by hawkfish** [comment]
Also, none of the `Tasks`s were being executed when I left out the `scheduler.GetTaskFromProducer` call. I assumed that the main thread was responsible for handling some of the work until it was all done?

**Comment by hawkfish** [comment]
Hmm, adding `pipeline.Finish()` to the end of the method seems to have fixed it. Is that safe?

**Comment by Mytherin** [comment]
> Yes I saw that but the problem is that there is no way to know ahead of time how many Tasks will be added because each (WPK, WOK) pair may generate completely different partitioning.

The code above seems like it knows how many tasks will be scheduled, no? `(matching.size() * ranges.size())`)

In that case it is similar to the hash aggregate, which also does not know how many tasks it needs to schedule until the partitioned sub-aggregates have completed. At scheduling time the total_tasks gets incremented by the number of tasks that will be scheduled, and those will then run.

It shouldn't even be a problem if those tasks then schedule additional tasks before finishing (as long as total_tasks gets incremented appropriately). 

> Also, none of the Tasks were being executed when I left out the scheduler.GetTaskFromProducer call. I assumed that the main thread was responsible for handling some of the work until it was all done?

> Also, none of the Tasks were being executed when I left out the scheduler.GetTaskFromProducer call. I assumed that the main thread was responsible for handling some of the work until it was all done?

That should be the case. I'm not sure why that is required. Are you scheduling tasks and testing  in the single-threaded case (i.e. without setting PRAGMA threads=X)? Perhaps there is an issue there. The partitioned aggregate HT only schedules tasks when there are multiple worker threads active.

> Hmm, adding pipeline.Finish() to the end of the method seems to have fixed it. Is that safe?

That should be fine as long as it is called only once. The Finish is called in the aggregate HT instead of the FinishTask, since the FinishTask also triggers a Finalize call to be triggered. In general I think this API should be reworked to better accommodate the multi-stage parallel operators, though.

**Comment by hawkfish** [comment]
It doesn't know how many tasks there will be because `matching` is only the functions that share a key pair. The next pass will end up with a completely different key pair, which will require repartitioning the tables - and we won't know how many tasks there are until after that (destructive) resorting.

As for not scheduling `Tasks in the single threaded case, I couldn't see how HT did that - it looked to me like it was always being called with `immediate` set to `false` (in `PhysicalHashAggregate::Finalize`) The only place I could see `PhysicalHashAggregate::FinalizeImmediate` being called was in `PhysicalDelimJoin::Finalize`.

**Comment by hawkfish** [comment]
The tests all run ;-)

**Comment by Mytherin** [comment]
> It doesn't know how many tasks there will be because matching is only the functions that share a key pair. The next pass will end up with a completely different key pair, which will require repartitioning the tables - and we won't know how many tasks there are until after that (destructive) resorting.

At that point you would schedule new tasks, and when you schedule new tasks you can increment this counter, no?

The counter is there to keep track of when the pipeline is completed; is there perhaps a better way of keeping track of this that doesn't involve counting the tasks?

> As for not scheduling `Tasks in the single threaded case, I couldn't see how HT did that - it looked to me like it was always being called with `immediate`set to`false`(in`PhysicalHashAggregate::Finalize`) The only place I could see `PhysicalHashAggregate::FinalizeImmediate`being called was in`PhysicalDelimJoin::Finalize`.

This happens in an earlier stage, actually. The Finalize is a nop in the case where there are no partitions, and the partitioning is prevented entirely in the case where there are no threads. Hence in the single-threaded case there will be no partitions and nothing to merge.

> The tests all run ;-)

Excellent, congrats!

**Comment by hawkfish** [comment]
I spoke too soon - the all unit segfaults.

**Comment by hawkfish** [comment]
I've added full parallelism, partitioning, bitfield boundaries and `OVER` clause grouping. Do we want to attack the sorting now? And if so, should we use the referenced block sorting or find some way to pull out the row-based sort from `PhysicalOrder`?

**Comment by lnkuiper** [comment]
I am still working on improving PhysicalOrder by adding parallelism and support for external sorting using the buffermanager. I'm happy to discuss if/how we can integrate (parts of) the PhysicalOrder in window functions @hawkfish , send me an e-mail at laurens.kuiper@cwi.nl if you'd like.

**Comment by hannesmuehleisen** [comment]
@stas-sl I believe this has been mostly mitigated thanks to @hawkfish . On my machine your example now runs in 0.3s . Closing this.
